File: test/language/optional_members/java/optional_members/OptionalArrayTest.java
package optional_members;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import optional_members.optional_array.TestStruct;
import optional_members.optional_array.Data8;

public class OptionalArrayTest
{
    @Test
    public void data8()
    {
        final TestStruct test = new TestStruct();
        test.setHasData8(true);
        test.setData8(new Data8[4]);
        assertEquals(4, test.getData8().length);
    }

    @Test
    public void autoData8()
    {
        final TestStruct test = new TestStruct();
        assertFalse(test.isAutoData8Used());
        test.setAutoData8(new Data8[4]);
        assertTrue(test.isAutoData8Used());
        assertEquals(4, test.getAutoData8().length);
    }

    @Test
    public void data16()
    {
        final TestStruct test = new TestStruct();
        test.setHasData8(false);
        test.setData16(new short[4]);
        assertEquals(4, test.getData16().length);
    }

    @Test
    public void autoData16()
    {
        final TestStruct test = new TestStruct();
        assertFalse(test.isAutoData16Used());
        test.setAutoData16(new short[4]);
        assertTrue(test.isAutoData16Used());
        assertEquals(4, test.getAutoData16().length);
    }
}

File: test/language/optional_members/cpp98/OptionalArrayTest.cpp
#include "gtest/gtest.h"

#include "optional_members/optional_array/TestStruct.h"

// just test setters and getters
namespace optional_members
{
namespace optional_array
{

TEST(OptionalArray, data8)
{
    TestStruct test;
    test.setHasData8(true);
    test.setData8(zserio::ObjectArray<Data8>(4));
    ASSERT_EQ(4, test.getData8().size());
}

TEST(OptionalArray, autoData8)
{
    TestStruct test;
    ASSERT_FALSE(test.hasAutoData8());
    test.setAutoData8(zserio::ObjectArray<Data8>(4));
    ASSERT_TRUE(test.hasAutoData8());
    ASSERT_EQ(4, test.getAutoData8().size());
}

TEST(OptionalArray, data16)
{
    TestStruct test;
    test.setHasData8(false);
    test.setData16(zserio::Int16Array(4));
    ASSERT_EQ(4, test.getData16().size());
}

TEST(OptionalArray, autoData16)
{
    TestStruct test;
    ASSERT_FALSE(test.hasAutoData16());
    test.setAutoData16(zserio::Int16Array(4));
    ASSERT_TRUE(test.hasAutoData16());
    ASSERT_EQ(4, test.getAutoData16().size());
}

} // namespace optional_array
} // namespace optional_members
